File: packages/iabtcf_consent_info/android/src/main/kotlin/com/terwesten/gabriel/iabtcf_consent_info/IabtcfConsentInfoPlugin.kt
package com.terwesten.gabriel.iabtcf_consent_info

import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import androidx.annotation.NonNull
import androidx.preference.PreferenceManager
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel

/** IabtcfConsentInfoPlugin */
class IabtcfConsentInfoPlugin : FlutterPlugin, EventChannel.StreamHandler {
    private lateinit var channel: EventChannel

    private lateinit var sharedPreferences: SharedPreferences

    private var events: EventChannel.EventSink? = null

    private var lastConsentInfoSent: Map<String, Any?>? = null

    private val sharedPreferencesListener = OnSharedPreferenceChangeListener { _, _ ->
        sendConsentInfo()
    }

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = EventChannel(flutterPluginBinding.binaryMessenger, "com.terwesten.gabriel/iabtcf_consent_info")
        channel.setStreamHandler(this)

        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(flutterPluginBinding.applicationContext)
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setStreamHandler(null)
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
        this.events = events
        sendConsentInfo()
        sharedPreferences.registerOnSharedPreferenceChangeListener(sharedPreferencesListener)
    }

    override fun onCancel(arguments: Any?) {
        sharedPreferences.unregisterOnSharedPreferenceChangeListener(sharedPreferencesListener)
        events = null
        lastConsentInfoSent = null
    }

    private fun sendConsentInfo() {
        val consentInfo = getConsentInfo()

        if (lastConsentInfoSent == null || consentInfo != lastConsentInfoSent) {
            lastConsentInfoSent = consentInfo
            events!!.success(consentInfo)
        }
    }

    private fun getConsentInfo(): Map<String, Any?> =
            sharedPreferences.all.filterKeys { it.startsWith("IABTCF_") }
}

File: packages/iabtcf_consent_info/ios/iabtcf_consent_info/Sources/iabtcf_consent_info/IabtcfConsentInfoPlugin.swift
import Flutter
import Foundation

public class IabtcfConsentInfoPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterEventChannel(
            name: "com.terwesten.gabriel/iabtcf_consent_info",
            binaryMessenger: registrar.messenger()
        )

        let instance = IabtcfConsentInfoPlugin()

        channel.setStreamHandler(instance)
    }

    private var events: FlutterEventSink!

    private var lastConsentInfoSent: [String: Any]?

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {

        self.events = events

        sendConsentInfo()

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(self.userDefaultsDidChange),
            name: UserDefaults.didChangeNotification,
            object: nil
        )

        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        NotificationCenter.default.removeObserver(
            self,
            name: UserDefaults.didChangeNotification,
            object: nil
        )

        lastConsentInfoSent = nil

        events = nil

        return nil
    }

    @objc
    public func userDefaultsDidChange() {
        sendConsentInfo()
    }

    func sendConsentInfo() {
        let consentInfo = getConsentInfo()

        if (lastConsentInfoSent == nil || !NSDictionary(dictionary: consentInfo).isEqual(to: lastConsentInfoSent!)) {
            lastConsentInfoSent = consentInfo
            events(consentInfo)
        }
    }

    private func getConsentInfo() -> [String: Any] {
        UserDefaults.standard.dictionaryRepresentation()
            .filter { $0.key.hasPrefix("IABTCF_")}
    }
}
